plugins/blueprint.cpp

```diff
@@ -592,179 +612,185 @@ command_result do_transform(DFCoord start, DFCoord end, string name, uint32_t ph
+    if (options.auto_phase || options.dig)
         dig.close();
+
     return CR_OK;
 }
 
-bool cmd_option_exists(vector<string>& parameters, const string& option)
+static bool get_options(blueprint_options &opts,
+                        const vector<string> &parameters)
 {
-    return  find(parameters.begin(), parameters.end(), option) != parameters.end();
+    auto L = Lua::Core::State;
+    color_ostream_proxy out(Core::getInstance().getConsole());
+    Lua::StackUnwinder top(L);
+
+    if (!lua_checkstack(L, parameters.size() + 2) ||
+        !Lua::PushModulePublic(
+            out, L, "plugins.blueprint", "parse_commandline"))
+    {
+        out.printerr("Failed to load blueprint Lua code\n");
+        return false;
+    }
+
+    Lua::Push(L, &opts);
+
+    for (const string &param : parameters)
+        Lua::Push(L, param);
+
+    if (!Lua::SafeCall(out, L, parameters.size() + 1, 0))
+        return false;
+
+    return true;
+}
+
+static void print_help()
+{
+    auto L = Lua::Core::State;
+    color_ostream_proxy out(Core::getInstance().getConsole());
+    Lua::StackUnwinder top(L);
+
+    if (!lua_checkstack(L, 1) ||
+        !Lua::PushModulePublic(out, L, "plugins.blueprint", "print_help") ||
+        !Lua::SafeCall(out, L, 0, 0))
+    {
+        out.printerr("Failed to load blueprint Lua code\n");
+    }
 }
 
 command_result blueprint(color_ostream &out, vector<string> &parameters)
 {
-    if (parameters.size() < 4 || parameters.size() > 8)
-        return help(out);
     CoreSuspender suspend;
-    if (!Maps::IsValid())
+
+    if (parameters.size() >= 1 && parameters[0] == "gui")
     {
-        out.printerr("Map is not available!\n");
-        return CR_FAILURE;
+        std::ostringstream command;
+        command << "gui/blueprint";
+        for (const string &param : parameters)
+        {
+            command << " " << param;
+        }
+        string command_str = command.str();
+        out.print("launching %s\n", command_str.c_str());
+
+        Core::getInstance().setHotkeyCmd(command_str);
+        return CR_OK;
     }
-    int32_t x, y, z;
-    if (!Gui::getCursorCoords(x, y, z))
+
+    blueprint_options options;
+    if (!get_options(options, parameters) || options.help)
     {
-        out.printerr("Can't get cursor coords! Make sure you have an active cursor in DF.\n");
+        print_help();
+        return options.help ? CR_OK : CR_FAILURE;
+    }
+
+    if (!Maps::IsValid())
+    {
+        out.printerr("Map is not available!\n");
         return CR_FAILURE;
     }
-    DFCoord start (x, y, z);
-    DFCoord end (x + stoi(parameters[0]), y + stoi(parameters[1]), z + stoi(parameters[2]));
-    uint32_t option = 0;
-    if (parameters.size() == 4)
+
+    // start coordinates can come from either the commandline or the map cursor
+    DFCoord start(options.start);
+    if (options.start.x == -30000)
     {
-        option = DIG | BUILD | PLACE | QUERY;
+        if (!Gui::getCursorCoords(options.start))
```

Comment:
I think both instances of `options.start` here should be `start`. As-is, `getCursorCoords()` is modifying `options.start`, but `start` is used later in the function. Specifically, the unmodified `start` fails the `isValidTilePos()` check below, so I always got the following error:
```
Invalid start position: -30000,-30000,-30000
```
until this change:
```diff
diff --git a/plugins/blueprint.cpp b/plugins/blueprint.cpp
index fc8ac5fa..2158f408 100644
--- a/plugins/blueprint.cpp
+++ b/plugins/blueprint.cpp
@@ -754,9 +754,9 @@ command_result blueprint(color_ostream &out, vector<string> &parameters)
 
     // start coordinates can come from either the commandline or the map cursor
     DFCoord start(options.start);
-    if (options.start.x == -30000)
+    if (start.x == -30000)
     {
-        if (!Gui::getCursorCoords(options.start))
+        if (!Gui::getCursorCoords(start))
         {
             out.printerr("Can't get cursor coords! Make sure you specify the"
                     " --cursor parameter or have an active cursor in DF.\n");
```

Reply:
thank you for catching this. obviously, I did not test enough after migrating to the new `getCursorCoords` API : /

done.